### PR TITLE
Improve `handlerTimeout`

### LINKED ISSUE #735: Long Polling drops after 1 day
* Telegraf.js Version: 3.32.0
* Node.js Version: v12.5.0
* Operating System: Ubuntu 18.04

Hi, I have a problem where the bot stops responding to any message after 1 day or so. I use long polling method because my server is in a corporate intranet with proxy connection to internet.  So webhook is no option for me. 

I use PM2 and the process is still running fine, the bot is still running. no errors from log.  I've been searching for answers for 3 days and haven't found it yet, so I'm writing here. please help..

### PR DESCRIPTION
**Pull request by wojpawlik**
# Description

Properly fixes #735. Closes #1247. It's nowhere near as sophisticated, but it's simple and leaks no implementation details to `telegraf.ts`, which should make it easy to improve over time if needed.

I used `Set` for storage for simplicity. Can be easily changed to other data structure later if _benchmark_ indicates it's faster.

All updates in one batch receive the same timer&mdash;when using webhook, earlier updates will have more time than `handlerTimeout`.

I can fix the notable omissions if you deem them important:
- No `catchTimeout`. It's not very useful without assigning `AbortSignal` to `ctx`, because this implementation still waits for all the updates to complete processing before polling for more updates.
- As a result, `Polling::syncUpdateOffset()` is not called on timeout.

## Type of change

- New feature
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Purposefully timing out on some, but not all received updates, using only long polling.

**Test Configuration**:
* Node.js Version: v14.0.0
* Operating System: Windows 2004 64-bit

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by KnorpelSenf** [comment]
This PR solves some of the same problems as #1247. I wrote down the parallels I noticed so we have a common understanding of what is going on:

#### Number of currently pending tasks (declaration, increment, decrement)

`DecayingDeque`
https://github.com/telegraf/telegraf/blob/f33c85a57e04623c6b61ec8115f19880f04824af/src/core/processing/queue.ts#L75
https://github.com/telegraf/telegraf/blob/f33c85a57e04623c6b61ec8115f19880f04824af/src/core/processing/queue.ts#L131
https://github.com/telegraf/telegraf/blob/f33c85a57e04623c6b61ec8115f19880f04824af/src/core/processing/queue.ts#L200

`Timeouts`
https://github.com/telegraf/telegraf/blob/569051cf186aea1277f5489965321250e4571916/src/core/timeouts.ts#L10
https://github.com/telegraf/telegraf/blob/569051cf186aea1277f5489965321250e4571916/src/telegraf.ts#L194
https://github.com/telegraf/telegraf/blob/569051cf186aea1277f5489965321250e4571916/src/telegraf.ts#L196

#### Handling completed middleware

`DecayingQueue`
https://github.com/telegraf/telegraf/blob/f33c85a57e04623c6b61ec8115f19880f04824af/src/core/processing/queue.ts#L228
`Timeouts`
https://github.com/telegraf/telegraf/blob/569051cf186aea1277f5489965321250e4571916/src/core/timeouts.ts#L20
and then
https://github.com/telegraf/telegraf/blob/569051cf186aea1277f5489965321250e4571916/src/telegraf.ts#L213
and
https://github.com/telegraf/telegraf/blob/569051cf186aea1277f5489965321250e4571916/src/telegraf.ts#L219

#### Other

`Drift`s exist in both implementations, and the `add` call does, too.

---

I fail to spot anything here that is not already included in #1247. If there are any advantages, could you outline them?

You use the terminology of drifts, but the elements do not drift, do they? Why the name, and is there any point to wrapping the context objects inside another object?

> leaks no implementation details

Could you explain where this happens in #1247? Even the backpressure mechanism for concurrent `getUpdates` calls is abstracted away by promises. I'm surprised about this comment.

When using `Timeouts`, `telegraf.ts` has to manually manage the number of currently processed updates. You're basically pulling this out of the timeout module and into `telegraf.ts`. Of course it makes the timeout module shorter to let other parts of the lib do the work, but why does this make sense?

The same is true for the `done` callback: #1247 sees it as a responsibility of the timeout handling module to remove tasks once they complete. You're suggesting here to take care of this in `telegraf.ts` instead. What is the point of managing this outside of the timeout handling module?

This module is not testable because of these two decisions, and untested code makes it hard to trust future changes.

It looks like the main reason why this implementation is shorter than #1247 is because it removes the ability to process updates concurrently across batches. I don't understand where to go from here if we wanted to implement that. Concurrently receiving updates happens when using webhooks.

Could you explain how this implementation behaves when used with webhooks, i.e. when adding contexts before earlier ones complete? It looks to me like we're starting a new timer for every update, and calling `unref` on it.

> Calling timeout.unref() creates an internal timer that will wake the Node.js event loop. Creating too many of these can adversely impact performance of the Node.js application. ([Source](https://nodejs.org/api/timers.html#timers_timeout_unref))

However, I'm not sure this is the intended behaviour, because it looks coincidental that `handleUpdate` and `handleUpdates` behave so differently. (One sets `minBatchSize` while the other one does not, even though `handleUpdate` is public API.)

What was the problem with #1247 that you started to rewrite it?

**Comment by wojpawlik** [comment]
#1247 was simply too complex for me to understand. I figured I'd either write something simpler, or find out why the complexity is unavoidable.

> You use the terminology of drifts, but the elements do not drift, do they? Why the name

I'm not sure what do you mean by "drifting", tbh. I used the name because you used it :)

> is there any point to wrapping the context objects inside another object?

Yes, it allows to add more properties (such as `promise` or `date`) whenever needed.

> Could you explain how this implementation behaves when used with webhooks, i.e. when adding contexts before earlier ones complete?

The ones added later will be grouped into artifical batch and receive a shared timeout.

**Comment by wojpawlik** [comment]
I'm feeling like coding right now, I'll review #1247 once again tomorrow.

**Comment by KnorpelSenf** [comment]
> The ones added later will be grouped into artificial batch and receive a shared timeout.

There is always only one update per webhook request. That does indeed mean we spawn a timer per update, which leads to as many concurrent timers as there are updates in 90 seconds (`handlerTimeout`).

---

> simply too complex for me to understand

Then I must have done a terrible job at explaining it despite all the lengthy comments :D

I'll try again from a different perspective, because it's really not much more complicated than what you wrote (it just has a few more lines).

> I'm not sure what do you mean by "drifting", tbh.

This is the natural behaviour of an element in a queue (any queue, really, not just deques). Let's look at this queue:

```
0-1-2-3-4
```
It has five elements. We now add `X`:
```
0-1-2-3-4-X
```
Once the other elements are removed, starting at the head, `X` drifts to the left:
```
0-1-2-3-4-X
1-2-3-4-X
2-3-4-X
3-4-X
4-X
X
```
Note that also `4` drifts to the left before it gets removed—in fact, this happens for all elements. That's why I called all elements drifts.

The difference for a decaying queue is that `0`-`4` are not simply popped from the head aka. removed from the left side. Remember that an element is removed once the middleware completes. This could happen at any time, so elements are in fact removed from anywhere in the queue, example:
```
0-1-2-3-4-X
0-1-3-4-X
0-3-4-X
0-3-X
3-X
X
```
The exact order depends on how long the middleware takes to complete.

Summing up, a drift is basically just called drift because it has that drifting property (it moves towards the head), and I needed a distinct name for the interface.

A decaying queue is not much more than an ordinary queue, but if you look at `0-1-2-3-4`, it is not popped strictly from the left as it is usual for queues. The elements are “popped” (removed) at arbitrary positions. I wanted to describe this weird creature (a queue that does not pop linearly), so I called this processes decaying.

**Comment by KnorpelSenf** [comment]
Just saw that now, thanks!

> I'm feeling like coding right now, I'll review #1247 once again tomorrow.

**Comment by wojpawlik** [comment]
You're right, I can do much better here, stay tuned!

I believe #1247's biggest mistake is `DecayingDeque` _being_ a queue instead of _containing_ one. I also dislike a lot of the changes in `telegraf.ts`.

**Comment by KnorpelSenf** [comment]
So you want two closely coupled files instead of one?

Or in your case, the `Set` being the second file that cannot be exchanged?

**Comment by wojpawlik** [comment]
These errors sound familiar...

This is about as good as it gets with `Set`, when I have more time I'll change to list to drop `Timeouts::minBatchSize`.

I have nothing against list anymore, I just want it to live in a separate file and have well-defined interface.

**Comment by KnorpelSenf** [comment]
I have an old version where the list was still in a separate file, but the code evolved since then. I'll redo it tomorrow for the current version of #1247. Maybe it's pretty now, maybe not, but either way we'll get something in our hands to make an informed decision.

While I'm at it, I can also revert the changes to `telegraf.ts`. Those were not strictly related to the queue. Maybe we can find something better, I also see room for improvement there.

**Comment by wojpawlik** [comment]
To be clear, `allowed_updates` for webhooks is a good fix, leave that in.

**Comment by wojpawlik** [comment]
Summary:

- moved to `Yallist`, `minBatchSize` no longer needed,
- still simpler and 43% shorter than #1247 / [`update-deque-separated`#diff-d6c763c542](https://github.com/telegraf/telegraf/compare/update-deque-separated#diff-d6c763c5428caa3fbe0ef778d7585d757be901c2ba548712a4f98ef6fd14cada),
- error handling not changed, so `done` is necessary to detect `handleError` timeouts,
- backpressure not changed 👍,
- default `handleTimeout` should probably `Polling::syncUpdateOffset`.

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
I started commenting on the major things. You made a number of odd decisions when it comes to modularization, see comments below.

Also, I don't see any trace of `concurrency` or backpressure the way #1247 does it. Did you push all commits? The number of updates to fetch with `getUpdates` never adjusts to the load.

**Comment by KnorpelSenf** [comment]
One more thing. You started out with this:

> I have nothing against list anymore, I just want it to live in a separate file and have well-defined interface.

Now you have neither. `telegraf.ts` takes care of handling timeouts, `Timeouts` takes care of handling a list, and you have an extra dependency that no one needs. The list is not in a separate file and we still don't have clear interfaces.



> I'm feeling like coding right now, I'll review #1247 once again tomorrow.

Did you find time to do this yet? #1247 does not suffer from any of these problems, and it has comments and tests. It certainly isn't perfect, though, that's why I need you to figure out if there's something wrong with it :)

**Comment by wojpawlik** [comment]
I looked at [`update-deque-separated`#diff-d6c763c542](https://github.com/telegraf/telegraf/compare/update-deque-separated#diff-d6c763c5428caa3fbe0ef778d7585d757be901c2ba548712a4f98ef6fd14cada) and I still think mine is simpler, so I'd rather work on this until we're both happy with it.

> `Timeouts` takes care of handling a list,

Kind of. It doesn't bother itself with shuffling references around, it leaves that to Yallist. And in the future, I may replace Yallist with a small list that has only stuff used by `Timeouts`.

> Also, I don't see any trace of `concurrency` or backpressure the way #1247 does it. Did you push all commits? The number of updates to fetch with `getUpdates` never adjusts to the load.

It's intentionally completely missing. Polling will work as it always has&mdash;request up to 100 updates, after current batch is processed.

File: src/core/timeouts.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  private isTimerRunning = false
+  private runTimer() {
+    const node = this.list.tail // head works too, sets timers more often
+    if (node == null || this.isTimerRunning) return
+    this.isTimerRunning = true
+    const timeLeft = node.value.timeoutsAt - Date.now()
+    const ms = Math.max(timeLeft, MIN_TIMEOUT_DURATION)
+    setTimeout(() => {
+      try {
+        while (true) {
```
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:45]: 1\. Why? 2. Should I use `for` loop?

File: src/core/timeouts.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import Yallist = require('yallist')
+
+interface Drift<C extends Context> {
+  readonly ctx: C
+  readonly timeoutsAt: number
+}
+
+const MIN_TIMEOUT_DURATION = 5_000 // 5s in ms
+
+export class Timeouts<C extends Context> {
```
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts]: Welp, I did it and it broke something, reverting, it can be dropped anytime later.

File: src/core/timeouts.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    )
+  }
+
+  private isTimerRunning = false
+  private runTimer() {
+    const node = this.list.tail // head works too, sets timers more often
+    if (node == null || this.isTimerRunning) return
+    this.isTimerRunning = true
+    const timeLeft = node.value.timeoutsAt - Date.now()
+    const ms = Math.max(timeLeft, MIN_TIMEOUT_DURATION)
```
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:42]: 1. Clearing timeouts is synchronous, no other work can happen in the meantime,
2. Scheduling a short timer is unnecessary work,
3. Users can pass negative timeouts.

This neatly addresses all 3.

> best-effort attempt to be as precise as possible

Changing this to `head` would have a bigger positive impact on accuracy: https://github.com/telegraf/telegraf/blob/c053ff3af9b097a91aa1888390c6059376deba6c/src/core/timeouts.ts#L38

File: src/telegraf.ts [hunk 4]
```diff
@@ -211,12 +216,18 @@ export class Telegraf<C extends Context = Context> extends Composer<C> {
...
+        ? this.timeouts.add({ ctx, timeoutsAt })
+        : anoop
     try {
       await this.middleware()(ctx, anoop)
     } catch (err) {
       return await this.handleError(err, ctx)
     } finally {
-      if (webhookResponse !== undefined && !webhookResponse.writableEnded) {
+      done()
+      if (webhookResponse?.writableEnded === false) {
```
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/telegraf.ts:230]: > 3. How come you have such a strict separation between different ways how middleware can fail? If it throws, you handle it here, if it times out, you don't. Why distribute handling failure across different modules?

Excellent point. I'm considering changing it to something akin to:

```ts
await this.timeouts.add(this.middleware(), ctx)
```

Promise returned by `.add` would either

- resolve if middleware succeeds,
- reject with `TimeoutError`,
- do whatever `handleError` does in case of error.

This would keep error handling straightforward; no matter what happens, all `finally` handlers will run.

What do you think?

File: src/core/timeouts.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  add(drift: Drift<C>) {
+    const node = new Yallist.Node(drift)
+    this.list.pushNode(node)
+    this.runTimer()
+    return () => {
+      if (node.list != null) {
+        this.list.removeNode(node)
+      }
+    }
```
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:23]: I never say I wouldn't take full advantage of list data structure. I only want to have reference updaing encapsulated.

**Comment by KnorpelSenf** [comment]
> I looked at [`update-deque-separated`#diff-d6c763c542](https://github.com/telegraf/telegraf/compare/update-deque-separated#diff-d6c763c5428caa3fbe0ef778d7585d757be901c2ba548712a4f98ef6fd14cada) and I still think mine is simpler

`update-deque-separated` is terrible. It has the same problems as your branch. #1252 may be simpler for you to understand because you wrote it. I didn't, and I find it more difficult already, especially because of all the blurry lines between modules.

> 
> > `Timeouts` takes care of handling a list,
>
> Kind of. It doesn't bother itself with shuffling references around, it leaves that to Yallist. And in the future, I may replace Yallist with a small list that has only stuff used by `Timeouts`.

So you're only moving around its objects without understanding how they're linked … that's worse.

> 
> > Also, I don't see any trace of `concurrency` or backpressure the way #1247 does it. Did you push all commits? The number of updates to fetch with `getUpdates` never adjusts to the load.
> 
> It's intentionally completely missing. Polling will work as it always has—request up to 100 updates, after current batch is processed.

This is a bummer, then `telegraf` still cannot be used in webhook mode with middleware that takes longer than a few seconds. If you don't want to use the feature for polling, you can leave it as it is, but implementing `concurrency` (or some alternative that does the same) is not optional. The Telegram servers resent webhook updates if you don't confirm them quickly enough. You have to be able to make `handleUpdate` resolve immediately (before middleware is done) while still catching both errors and timeouts, otherwise `telegraf` will process the same updates again and again and eventually become unusable (been there, done that). If you omit solving this problem, `telegraf` consumers either have to roll their own update deduplication logic, or set up their own task queues (basically reimplement the relevant parts of #1247 in custom code).

File: src/telegraf.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/telegraf.ts:230]: Let's merge #1258 if you like it, before I create conflicts.

[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/telegraf.ts:230]: Better idea: remove `handleTimeout` entirely, signal them as passing a `TimeoutError` to `handleError`!

**Comment by wojpawlik** [comment]
> The Telegram servers resent webhook updates if you don't confirm them quickly enough

Even if bot doesn't close the connection?

File: src/core/timeouts.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:45]: I'll leave as-is for now, I'm changing other things right now

File: src/telegraf.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/telegraf.ts:230]: Almost ready, one type error to sort out, I can push if you want.

File: src/core/timeouts.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply src/core/timeouts.ts:23]: That makes no sense at all. It is arbitrary.

File: src/core/timeouts.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply src/core/timeouts.ts:42]: > 1\. Clearing timeouts is synchronous, no other work can happen in the meantime,

How does that relate to the problem? Being synchronous is no benefit if you're 5 seconds late

> 2\. Scheduling a short timer is unnecessary work,

That makes it a premature performance optimization that is not backed by runtime analyses or performance tests, so it bad to trade in reliability for it.

> 3\. Users can pass negative timeouts.

Try to explain that to people. “Timeout `-1000` means that the middleware has to be completed one second before the update arrives.”

> Changing this to `head`

It is `(head === null) === (tail === null)` (unless something is messed up). If there is a difference here, something is broken.

File: src/core/timeouts.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply src/core/timeouts.ts:45]: You have to be extremely cautious about changing anything, otherwise you may end up with an infinite loop that is very hard to debug. Move the condition to the loop is always safer and easier to read (no need for a flow analysis implemented in brainware).

**Comment by KnorpelSenf** [comment]
> > The Telegram servers resent webhook updates if you don't confirm them quickly enough
> 
> Even if bot doesn't close the connection?

Yes. (Webhook support by Telegram is a problem in itself.)

I wish we could make this cleaner, but it is the only thing I can come up with if we want to deal with the situation.

File: src/telegraf.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply src/telegraf.ts:230]: Not quite sure how it will look then. Also not sure if it's possible to fix it without refactoring `Timeouts`, hence also not sure if I want to invest more time in this PR

[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply src/telegraf.ts:230]: Anything I say here is pointing out how #1247 solves things better than this branch. We're reinventing the wheel here, because you don't want to read the ready solution?

**Comment by wojpawlik** [comment]
Perhaps `webhookReply` isn't just a premature optimization after all...

File: src/core/timeouts.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:40]: > It is `(head === null) === (tail === null)` (unless something is messed up). If there is a difference here, something is broken.

Yes, but `timeoutsAt` might differ.

[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:46]: If there's multiple updates with different `timeoutsAt`,

- `const node = this.list.head` will set the timer for the first, later updates will get a separate timer,
- `const node = this.list.tail` will set the timer for the last, earlier updates get more time.

About `MIN_TIMEOUT_DURATION`, I fail to see how setting short timers one after the other would do any good for performance. I'm willing to lower it to 1000 ms tho.

We cannot guarantee accuracy anyway due to single execution thread in JS, CPU scheduling, `SIGSTOP`... we can only guarantee we won't time out early. Also, the whole purpose of this module is to set timers more sparingly that naively racing each promise.

[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:46]: You're right. This is a premature optimization. I'll remove it and add some logging and validation instead.

File: src/telegraf.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/telegraf.ts]: I genuinely think my code is superior because of its simplicity. #1247 is concerning to me. Tomorrow I'll sit down and review it thoroughly to explain why.

File: src/core/timeouts.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wojpawlik) [reply src/core/timeouts.ts:46]: I didn't realize how long 90 seconds is :) let's avoid giving tasks more time than that.

**Comment by wojpawlik** [comment]
I wasn't aware of the issues with webhook, so I deem it out of scope. As a quickfix, `ApiClient::response` could be changed from `private readonly` to `readonly`, so that consumers can manually `ctx.tg.response.end()` if needed.

**Comment by wojpawlik** [comment]
Closing in favor of simpler #1262
